src/main/java/com/mraof/minestuck/network/EditmodeDragPacket.java

```diff
@@ -0,0 +1,281 @@
+package com.mraof.minestuck.network;
+
+import com.mraof.minestuck.MinestuckConfig;
+import com.mraof.minestuck.alchemy.*;
+import com.mraof.minestuck.computer.editmode.*;
+import com.mraof.minestuck.player.PlayerSavedData;
+import com.mraof.minestuck.skaianet.SburbConnection;
+import com.mraof.minestuck.util.MSCapabilities;
+import com.mraof.minestuck.util.MSSoundEvents;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Direction;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.sounds.SoundSource;
+import net.minecraft.util.Tuple;
+import net.minecraft.world.InteractionHand;
+import net.minecraft.world.InteractionResult;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.item.BlockItem;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.context.UseOnContext;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.SoundType;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.gameevent.GameEvent;
+import net.minecraft.world.phys.BlockHitResult;
+import net.minecraft.world.phys.Vec3;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public final class EditmodeDragPacket
+{
+	private static final Logger LOGGER = LogManager.getLogger();
+	
+	private static Tuple<Boolean, GristSet> editModePlaceCheck(EditData data, Player player, InteractionHand hand, BlockPos pos)
+	{
+		SburbConnection connection = data.getConnection();
+		ItemStack stack = player.getMainHandItem();
+		
+		if(stack.isEmpty() || hand.equals(InteractionHand.OFF_HAND) || !player.level.getBlockState(pos).getMaterial().isReplaceable())
+			return new Tuple<>(false, new GristSet());
+
+		DeployEntry entry = DeployList.getEntryForItem(stack, connection, player.level);
+		GristSet cost = entry != null ? entry.getCurrentCost(connection) : GristCost.findCostForItem(stack, null, false, player.level);
+		if(!GristHelper.canAfford(PlayerSavedData.getData(connection.getClientIdentifier(), player.level).getGristCache(), cost))
```

Comment:
Where does the cost during editmode fill actually get taken? My first instinct is that it would be in association with checking whether it can be afforded (to ensure that the same cost is used), but that does not appear to be the case.

Reply:
It gets taken in the server-side block break event, because breaking the block runs that event.

Reply:
That's something I would want changed for less messy code then. However since it's no simple change, I think that could wait until a PR after 4/13 to avoid introducing new bugs or take development time away from other things.

Reply:
I was already planning to completely rewrite the tool code at some point after 4/13 anyway, since it's not particularly modular. I think making turning them into a state machine of sorts would work well. But as you said, that's for after 4/13.